examples/vision/fully_convolutional_network.py

```diff
@@ -0,0 +1,630 @@
+
+We visualize what a random sample in our test split of the dataset looks like, and plot
+the segmentation mask on top to see the effective mask areas.
+Note that we have performed pre-processing on this dataset too,
+which makes the image and mask size same.
+"""
+
+# Select random image and mask. Cast to NumPy array
+# for Matplotlib visualization.
+
+images, masks = next(iter(test_ds))
+random_idx = tf.random.uniform([], minval=0, maxval=BATCH_SIZE, dtype=tf.int32)
+
+test_image = images[random_idx].numpy().astype("float")
+test_mask = masks[random_idx].numpy().astype("float")
+
+# Overlay segmentation mask on top of image.
+fig, ax = plt.subplots(nrows=1, ncols=2, figsize=(10, 5))
+
+ax[0].set_title("Image")
+ax[0].imshow(test_image / 255.0)
+
+ax[1].set_title("Image with segmentation mask overlay")
+ax[1].imshow(test_image / 255.0)
+ax[1].imshow(
+    test_mask,
+    cmap="inferno",
+    alpha=0.6,
+)
+plt.show()
+
+"""
+## Perform VGG-specific pre-processing
+
+`keras.applications.VGG19` requires the use of a `preprocess_input` function that will
+pro-actively perform Image-net style Standard Deviation Normalization scheme.
+"""
+
+
+def preprocess_data(image, segmentation_mask):
+    image = keras.applications.vgg19.preprocess_input(image)
+
+    return image, segmentation_mask
+
+
+train_ds = (
+    train_ds.map(preprocess_data, num_parallel_calls=AUTOTUNE)
+    .shuffle(buffer_size=1024)
+    .prefetch(buffer_size=1024)
+)
+valid_ds = (
+    valid_ds.map(preprocess_data, num_parallel_calls=AUTOTUNE)
+    .shuffle(buffer_size=1024)
+    .prefetch(buffer_size=1024)
+)
+test_ds = (
+    test_ds.map(preprocess_data, num_parallel_calls=AUTOTUNE)
+    .shuffle(buffer_size=1024)
+    .prefetch(buffer_size=1024)
+)
+"""
+## Model Definition
+
+The Fully-Convolutional Network boasts a simple architecture composed of only
+`keras.layers.Conv2D` Layers, `keras.layers.Dense` layers and `keras.layers.Dropout`
+layers.
+| ![FCN Architecture](https://i.imgur.com/PerTKjf.png) |
+| :--: |
+| **Diagram 2**: Generic FCN Forward Pass (Source: Paper)|
+
+Pixel-wise prediction is performed by having a Softmax Convolutional layer with the same
+size of the image, such that we can perform direct comparison
+We can find several important metrics such as Accuracy and Mean-Intersection-over-Union on the network.
+"""
+
+"""
+### Backbone (VGG-19)
+
+We use the [VGG-19 network](https://keras.io/api/applications/vgg/) as the backbone, as
+the paper suggests it to be one of the most effective backbones for this network.
+We extract different outputs from the network by making use of `keras.models.Model`.
+Following this, we add layers on top to make a network perfectly simulating that of
+Diagram 1.
+The backbone's `keras.layers.Dense` layers will be converted to `keras.layers.Conv2D`
+layers based on the [original Caffe code present
+here.](https://github.com/linxi159/FCN-caffe/blob/master/pascalcontext-fcn16s/net.py)
```

Comment:
Don't break links -- all links should be on a single line.

Reply:
Done!